pkg/utils/provider_meta.go

```diff
@@ -0,0 +1,94 @@
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/MaterializeInc/terraform-provider-materialize/pkg/clients"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/jmoiron/sqlx"
+)
+
+type ProviderMeta struct {
+	DB             map[clients.Region]*clients.DBClient
+	Frontegg       *clients.FronteggClient
+	CloudAPI       *clients.CloudAPIClient
+	DefaultRegion  clients.Region
+	RegionsEnabled map[clients.Region]bool
+}
+
+var DefaultRegion string
+
+func GetProviderMeta(meta interface{}) (*ProviderMeta, bool) {
+	providerMeta, ok := meta.(*ProviderMeta)
+	if !ok || providerMeta == nil {
+		fmt.Println("Type assertion failed: provider meta is not of type *ProviderMeta or is nil")
+		return nil, false
+	}
+	if providerMeta.Frontegg.NeedsTokenRefresh() {
+		err := providerMeta.Frontegg.RefreshToken()
+		if err != nil {
+			fmt.Printf("Failed to refresh token: %v\n", err)
```

Comment:
Delete?

Reply:
Why does this function return a `bool` instead of an `error`? Seems like this just wants to return the error in this case, and then the caller can wrap the error with more context.

Reply:
Good catch! Just changed this to actually return an error so we could use the standard:

```golang
if err != nil {
	return diag.FromErr(err)
}
```